### PR TITLE
introduce `make help` as default make recipe

### LINKED ISSUE #3769: Use code generation instead of `eval` in makefile
Also relates to https://github.com/chaos-mesh/chaos-mesh/issues/1963

Currently, we are using `eval` for rendering targets for "building executable binary" and "building container" images:

https://github.com/chaos-mesh/chaos-mesh/blob/cfbca5e5becfb09b0129a8d9702a930942999730/Makefile#L172-L175

https://github.com/chaos-mesh/chaos-mesh/blob/cfbca5e5becfb09b0129a8d9702a930942999730/Makefile#L211-L220

It's not easy to understand and maintain that makefile with dynamic rendering with `eval`. Also, it would reduce the performance of the bash completion for the makefile.

I suggest writing a go utility application, and generating `binary.generated.mk` and `container-image.generated.mk`, instead of using `eval` in `Makefile`

### PR DESCRIPTION
**Pull request by STRRL**
### What problem does this PR solve?

close #3769

### What's changed and how it works?

In this PR, we provide a target called `make help` like this:

image

### Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`
- Need to **cheery-pick to release branches**
  - [ ] release-2.5
  - [ ] release-2.4

### Checklist

CHANGELOG

- [x] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] **Breaking backward compatibility**

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

**Comment by STRRL** [comment]
/cc @g1eny0ung 

PTAL

**Comment by g1eny0ung** [comment]
I will look at it this afternoon and this evening.

File: hack/local-up-chaos-mesh.sh
Comment (g1eny0ung) [inline hack/local-up-chaos-mesh.sh:110]:
```suggestion
    IMAGE_TAG=$IMAGE_TAG UI=1 SWAGGER=1 make image
```

If I understand correctly, now we are no longer providing the ability to customize the registry, and some targets with `docker-push` prefix also are being removed.

I agree with these changes, since it's easy to rename the image name with `docker image tag`, we don't need to keep these seemingly convenient, but somewhat redundant targets.

So the `IMAGE_REGISTRY_PREFIX` env should be removed here as well as others.

File: cmd/generate-makefile/generate_container_image_makefile.go
Comment (g1eny0ung) [inline cmd/generate-makefile/generate_container_image_makefile.go:46]:
```suggestion
image-{{ .ImageName }}: {{ .SourcePath }}/.dockerbuilt ## {{ .Comment }}
```

**Review by g1eny0ung — COMMENTED** [review]
The rest LGTM. As it's too late, I'll check it again this morning.

File: cmd/generate-makefile/funcmaps.go
Comment (g1eny0ung) [inline cmd/generate-makefile/funcmaps.go:22]: It is better to check for empty strings.

File: Makefile
Comment (STRRL) [reply Makefile:74]: Great! Any reason for using `21` instead of `15`, not even longer? Or should use a larger number here?

File: hack/local-up-chaos-mesh.sh
Comment (STRRL) [reply hack/local-up-chaos-mesh.sh:110]: > If I understand correctly, now we are no longer providing the ability to customize the registry, and some targets with `docker-push` prefix also are being removed.

Yeah, that's the thing I want to do: do NOT provide the flags/ENVs for configuring different registry/repo, but **only during the building process**.

It seems the changes in `local-up-chaos-mesh.sh` is NOT a part of building things but installing things. 

So maybe it's better to revert the changes in `local-up-chaos-mesh.sh`.

If necessary,  I would open another PR for similar changes, but for installing scripts like `install.sh`, `local-up-chaos-mesh.sh` and other .sh scripts.

What do you think about it? Reverting changes in `local-up-chaos-mesh.sh`

File: cmd/generate-makefile/generate_container_image_makefile.go
Comment (STRRL) [reply cmd/generate-makefile/generate_container_image_makefile.go:46]: Nice Catch!

File: cmd/generate-makefile/funcmaps.go
Comment (STRRL) [reply cmd/generate-makefile/funcmaps.go:22]: `strings.Join` already takes care about that :P

https://github.com/golang/go/blob/e8fe3b77574d9afadab664104a669d8e9f0f4303/src/strings/strings.go#L432-L464

Comment (STRRL) [reply cmd/generate-makefile/funcmaps.go:22]: I would mark this comment as resovled

File: hack/local-up-chaos-mesh.sh
Comment (g1eny0ung) [reply hack/local-up-chaos-mesh.sh:110]: > What do you think about it? Reverting changes in local-up-chaos-mesh.sh

I still don't think it's necessary because these scripts are used more for debugging purposes, and giving them more configurable items is often redundant. As I said above:

_I agree with these changes, since it's easy to rename the image name with `docker image tag`, we don't need to keep these seemingly convenient, but somewhat redundant targets._

Comment (g1eny0ung) [reply hack/local-up-chaos-mesh.sh:110]: @STRRL Only this place is not updated, after the update I will approve this PR.

**Comment by cwen0** [comment]
Cool! I have run this pr on the environment, The `make help` is very useful!

File: Makefile
Comment (STRRL) [reply Makefile]: got that.

File: hack/local-up-chaos-mesh.sh
Comment (STRRL) [reply hack/local-up-chaos-mesh.sh:110]: Hi @g1eny0ung , I actually prefer to revert changes in `local-up-chaos-mesh.sh` **in this PR**.

I totally agree with your opinion. But it's a little out of the scope from this PR.

I would do it, but in another new PR. Not carry changes in `.sh` files in the PR.

What do you think about it?

Comment (g1eny0ung) [reply hack/local-up-chaos-mesh.sh:110]: > But it's a little out of the scope from this PR.

@STRRL That sounds reasonable. I agree that we can wrap up the `.sh` modifications in the upcoming PR.

**Review by g1eny0ung — APPROVED** [review]
LGTM!

**Comment by g1eny0ung** [comment]
@chaos-mesh/maintainers @chaos-mesh/committers Please be aware that any additions or removals of make targets in the `Makefile` must now be done with consideration for modifying the template.

**Pull request merged**
